octodns/provider/azuredns.py

```diff
@@ -0,0 +1,369 @@
+            :param record: An OctoDNS record
+            :type  record: ..record.Record
+            :param values: Parameters for a record. eg IP address, port, domain
+                           name, etc. Values usually read from record.data
+            :type  values: {'values': [...]} or {'value': [...]}
+
+            :type return: _AzureRecord
+        '''
+        self.resource_group = resource_group
+        self.zone_name = record.zone.name[0:len(record.zone.name) - 1]
+        self.relative_record_set_name = record.name or '@'
+        self.record_type = record._type
+
+        if delete:
+            return
+
+        format_u_s = '' if record._type == 'A' else '_'
+        key_name = '{}{}records'.format(self.record_type, format_u_s).lower()
+        class_name = '{}'.format(self.record_type).capitalize() + \
+                     'Record'.format(self.record_type)
+
+        self.params = getattr(self, '_params_for_{}'.format(record._type))
+        self.params = self.params(record.data, key_name, eval(class_name))
+        self.params['ttl'] = record.ttl
+
+    def _params(self, data, key_name, azure_class):
+        return {key_name: [azure_class(v) for v in data['values']]} \
+            if 'values' in data else {key_name: [azure_class(data['value'])]}
+
+    _params_for_A = _params
+    _params_for_AAAA = _params
+    _params_for_NS = _params
+    _params_for_PTR = _params
+    _params_for_TXT = _params
+
+    def _params_for_SRV(self, data, key_name, azure_class):
+        params = []
+        if 'values' in data:
+            for vals in data['values']:
+                params.append(azure_class(vals['priority'],
+                                          vals['weight'],
+                                          vals['port'],
+                                          vals['target']))
+        else:  # single value at key 'value'
+            params.append(azure_class(data['value']['priority'],
+                                      data['value']['weight'],
+                                      data['value']['port'],
+                                      data['value']['target']))
+        return {key_name: params}
+
+    def _params_for_MX(self, data, key_name, azure_class):
+        params = []
+        if 'values' in data:
+            for vals in data['values']:
+                params.append(azure_class(vals['preference'],
+                                          vals['exchange']))
+        else:  # single value at key 'value'
+            params.append(azure_class(data['value']['preference'],
+                                      data['value']['exchange']))
+        return {key_name: params}
+
+    def _params_for_CNAME(self, data, key_name, azure_class):
+        return {'cname_record': CnameRecord(data['value'])}
+
+    def _equals(self, b):
+        '''Checks whether two records are equal by comparing all fields.
+            :param b: Another _AzureRecord object
+            :type  b: _AzureRecord
```

Comment:
Would be nice to make a full pass of doing formal `__doc__` like this across all of octoDNS, ... someday.